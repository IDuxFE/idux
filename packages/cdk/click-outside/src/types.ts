/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type {
  ExtractInnerPropTypes,
  ExtractPublicPropTypes,
  MaybeArray,
  MaybeElement,
  MaybeElementRef,
} from '@idux/cdk/utils'
import type { Component, DefineComponent, HTMLAttributes, ObjectDirective, PropType } from 'vue'

export const clickOutsideProps = {
  disabled: { type: Boolean, default: false },
  is: { type: [String, Object] as PropType<string | Component>, default: 'div' },
  options: { type: Object as PropType<ClickOutsideOptions>, default: undefined },
  onClickOutside: [Function, Array] as PropType<MaybeArray<(evt: MouseEvent) => void>>,
} as const

export type ClickOutsideProps = ExtractInnerPropTypes<typeof clickOutsideProps>
export type ClickOutsidePublicProps = ExtractPublicPropTypes<typeof clickOutsideProps>
export type ClickOutsideComponent = DefineComponent<
  Omit<HTMLAttributes, keyof ClickOutsidePublicProps> & ClickOutsidePublicProps
>
export type ClickOutsideInstance = InstanceType<DefineComponent<ClickOutsideProps>>

export type ClickOutsideDirective = ObjectDirective<MaybeElement, ClickOutsideBinding>

export interface ClickOutsideOptions {
  container?: MaybeElementRef | Window | Document
  exclude?: MaybeElementRef[]
}
export type ClickOutsideHandler = (evt: MouseEvent) => void
export type ClickOutsideBinding = ClickOutsideHandler | (ClickOutsideOptions & { handler: ClickOutsideHandler })
