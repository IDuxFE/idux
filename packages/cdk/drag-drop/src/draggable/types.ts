/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import type { ExtractInnerPropTypes, ExtractPublicPropTypes, MaybeArray } from '@idux/cdk/utils'
import type { Component, DefineComponent, HTMLAttributes, PropType } from 'vue'

export const draggableProps = {
  disabled: { type: Boolean, default: false },
  is: { type: [String, Object] as PropType<string | Component>, default: 'div' },
  onStart: [Function, Array] as PropType<MaybeArray<DraggableEvent>>,
  onMove: [Function, Array] as PropType<MaybeArray<DraggableEvent>>,
  onEnd: [Function, Array] as PropType<MaybeArray<DraggableEvent>>,
} as const

export type DraggableProps = ExtractInnerPropTypes<typeof draggableProps>
export type DraggablePublicProps = ExtractPublicPropTypes<typeof draggableProps>
export type DraggableComponent = DefineComponent<
  Omit<HTMLAttributes, keyof DraggablePublicProps> & DraggablePublicProps
>
export type DraggableInstance = InstanceType<DefineComponent<DraggableProps>>

export interface DragPosition {
  left: number
  top: number
  offsetX: number
  offsetY: number
}

export type DraggableEvent = (position: DragPosition, evt: PointerEvent) => void

export interface DraggableOptions {
  onStart?: DraggableEvent
  onMove?: DraggableEvent
  onEnd?: DraggableEvent
}
