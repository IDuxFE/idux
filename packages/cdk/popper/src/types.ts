/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { Modifier, PositioningStrategy, State } from '@popperjs/core'
import type { ComponentPublicInstance, ComputedRef, Ref } from 'vue'

export type PopperPositioningStrategy = PositioningStrategy
export type PopperTrigger = 'click' | 'hover' | 'focus' | 'contextmenu' | 'manual'
export type PopperElement = ComponentPublicInstance | HTMLElement
export type PopperPlacement =
  | 'topStart'
  | 'top'
  | 'topEnd'
  | 'rightStart'
  | 'right'
  | 'rightEnd'
  | 'bottomStart'
  | 'bottom'
  | 'bottomEnd'
  | 'leftStart'
  | 'left'
  | 'leftEnd'

export interface PopperTriggerEvents {
  onClick?(event: Event): void
  onMouseenter?(event: Event): void
  onMouseleave?(event: Event): void
  onFocus?(event: Event): void
  onBlur?(event: Event): void
  onContextmenu?(event: Event): void
}

export interface PopperEvents {
  onMouseenter?(event: Event): void
  onMouseleave?(event: Event): void
}

export interface PopperOptions {
  /**
   * Whether to allow the mouse to enter the popper
   * * default is `true`
   */
  allowEnter?: boolean
  /**
   * Whether auto adjust when space is not enough
   * * default is `true`
   */
  autoAdjust?: boolean
  /**
   * Delay in ms once a trigger event is fired before a tippy shows or hides
   *
   * @example
   * * 0: disable delay, it is default value
   * * 100: show and hide delay are 100ms
   * * [100, 200]: show delay is 100ms, hide delay is 200ms
   * * [100, null]: show delay is 100ms, hide delay is the default
   */
  delay?: number | [number | null, number | null]
  /**
   * Disable the popper
   * * default is `false`
   */
  disabled?: boolean
  /**
   * Popper offset.
   * * [Horizontal axis offset, vertical axis offset]
   * * default is `[0, 0]`
   */
  offset?: [number, number]
  /**
   * Alignment of popper
   * * default is `top`
   */
  placement?: PopperPlacement
  /**
   * Trigger method
   * * default is `hover`
   */
  trigger?: PopperTrigger
  /**
   * Control the visibility of the popper
   * * default is `false`
   */
  visible?: boolean

  strategy?: PopperPositioningStrategy
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  modifiers?: Array<Partial<Modifier<unknown, any>>>
  onFirstUpdate?: (state: Partial<State>) => void
}

export interface PopperInstance<TE extends PopperElement = PopperElement, PE extends PopperElement = PopperElement> {
  /**
   * Initialize the popper.
   * The life cycle of the popper will enter mounted.
   */
  initialize(): void
  /**
   * Show the popper.
   * The style of the popper container will be set to block.
   */
  show(delay?: number): void
  /**
   * Hide the popper.
   * The style of the popper container will be set to none.
   */
  hide(delay?: number): void
  /**
   * Update popper.
   */
  update(options: Partial<PopperOptions>): void
  /**
   * Force update popper
   */
  forceUpdate(): void
  /**
   * Destroy the popper.
   * The life cycle of the popper will enter beforeDestroy.
   * To show the popper again, please recreate.
   */
  destroy(): void
  /**
   * The display status of the current popper.
   * Control by visible and disable.
   */
  visibility: ComputedRef<boolean>
  /**
   * The current true position of the popper.
   */
  placement: ComputedRef<PopperPlacement>
  /**
   * The truth DOM node of the trigger.
   * The caller needs to bind the variable to the view.
   */
  triggerRef: Ref<TE | undefined>
  /**
   * Manually bind to the event on the trigger.
   */
  triggerEvents: ComputedRef<PopperTriggerEvents>
  /**
   * The truth DOM node of the popper.
   * The caller needs to bind the variable to the view.
   */
  popperRef: Ref<PE | undefined>
  /**
   * Manually bind to events on the popper.
   */
  popperEvents: ComputedRef<PopperEvents>
  /**
   * The truth DOM node of the arrow.
   * The caller needs to bind the variable to the view.
   */
  arrowRef: Ref<HTMLElement | undefined>
}
