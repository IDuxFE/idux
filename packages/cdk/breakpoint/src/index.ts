export * from './observe'
export * from './breakpoints'
export * from './mediaMatcher'
