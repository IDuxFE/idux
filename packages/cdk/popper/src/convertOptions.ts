/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { kebabCase } from 'lodash-es'

import {
  type ComputePositionConfig,
  type Middleware,
  type Placement,
  flip,
  offset as offsetMiddleware,
  shift,
} from '@floating-ui/dom'

import { type BaseOptions } from './composables/useOptions'
import { arrow } from './middlewares/arrow'
import { referenceHidden } from './middlewares/refenceHidden'
import { updatePlacement } from './middlewares/updatePlacement'
import { type PopperPlacement } from './types'

export interface ExtraOptions {
  arrowElement: HTMLElement | undefined
  updatePlacement: (value: PopperPlacement) => void
}

export function convertOptions(baseOptions: BaseOptions, extraOptions: ExtraOptions): ComputePositionConfig {
  const { placement, strategy, middlewares, offset, autoAdjust } = baseOptions
  const { arrowElement, updatePlacement: _updatePlacement } = extraOptions

  return {
    placement: kebabCase(placement) as Placement,
    strategy,
    middleware: [
      !!arrowElement && arrow(arrowElement),
      offsetMiddleware({
        mainAxis: offset[1],
        crossAxis: offset[0],
      }),
      autoAdjust && flip({ padding: 2, crossAxis: false }),
      autoAdjust && shift(),
      referenceHidden(),
      updatePlacement(_updatePlacement),
      ...middlewares,
    ].filter(Boolean) as Middleware[],
  }
}
