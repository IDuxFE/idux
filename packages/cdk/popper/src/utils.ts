import type { Ref } from 'vue'
import type { Options, Placement } from '@popperjs/core'
import type { BaseOptions } from './hooks'
import type { PopperElement, PopperPlacement } from './types'

import { unref } from 'vue'
import { camelCase, kebabCase } from 'lodash'
import { isHTMLElement } from '@idux/cdk/utils'

export function convertElement(elementRef: Ref<PopperElement | null> | PopperElement | null): HTMLElement | null {
  const element = unref(elementRef)
  if (!element) {
    return null
  }
  return isHTMLElement(element) ? element : element.$el
}

export interface ExtraOptions {
  arrowElement: HTMLElement | null
  updatePlacement: (value: PopperPlacement) => void
}

export function convertOptions(baseOptions: BaseOptions, extraOptions: ExtraOptions): Options {
  const { placement, strategy, onFirstUpdate, modifiers, offset, autoAdjust } = baseOptions
  const { arrowElement, updatePlacement } = extraOptions
  return {
    placement: kebabCase(placement) as Placement,
    strategy,
    onFirstUpdate,
    modifiers: [
      { name: 'offset', options: { offset } },
      { name: 'flip', enabled: autoAdjust, options: { padding: 4 } },
      { name: 'arrow', enabled: !!arrowElement, options: { element: arrowElement, padding: 4 } },
      {
        name: 'IDUX_updatePlacement',
        enabled: true,
        phase: 'beforeWrite',
        requires: ['computeStyles'],
        fn: ({ state }) => updatePlacement(camelCase(state.placement) as PopperPlacement),
      },
      ...modifiers,
    ],
  }
}
