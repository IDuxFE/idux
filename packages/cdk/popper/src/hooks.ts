import type { ComputedRef, Ref, WatchStopHandle } from 'vue'
import type { Placement } from '@popperjs/core'
import type {
  PopperElement,
  PopperEvents,
  PopperOptions,
  PopperPlacement,
  PopperTrigger,
  PopperTriggerEvents,
} from './types'

import { computed, reactive, ref, watch } from 'vue'
import { camelCase, kebabCase } from 'lodash'
import { isUndefined } from '@idux/cdk/utils'
import { mapTriggerEvents } from '@idux/cdk/popper/src/utils'

export type PopperState = Required<Omit<PopperOptions, 'placement'> & { placement: Placement }>

export function useState(options: PopperOptions): PopperState {
  const state = reactive<Required<Omit<PopperOptions, 'placement'> & { placement: Placement }>>({
    visible: true,
    scrollStrategy: 'reposition',
    disabled: false,
    placement: 'top',
    trigger: 'manual',
    allowEnter: false,
    autoAdjust: true,
    offset: [0, 0],
    hideDelay: 0,
    showDelay: 0,
  })

  watch(
    () => state.placement,
    value => {
      state.placement = kebabCase(value) as Placement
    },
  )

  for (const [key, value] of Object.entries(options)) {
    if (isUndefined(value)) {
      continue
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    state[key] = value
  }

  return state
}

export function useVisibility(state: PopperState, action: () => void): [ComputedRef<boolean>, WatchStopHandle] {
  const visibility = computed(() => !state.disabled && state.visible)

  const watcher = watch(
    visibility,
    value => {
      if (value) {
        action()
      }
    },
    { flush: 'post' },
  )

  return [visibility, watcher]
}

export function usePlacement(state: PopperState, action: () => void): [ComputedRef<PopperPlacement>, WatchStopHandle] {
  const placement = computed(() => camelCase(state.placement) as PopperPlacement)

  const watcher = watch(
    placement,
    () => {
      action()
    },
    { flush: 'post' },
  )

  return [placement, watcher]
}

export function useElement<T extends PopperElement>(): Ref<T | null> {
  const element: Ref<T | null> = ref(null)
  return element
}

export function useTimer(): [Ref<NodeJS.Timer | null>, (fn: () => void, delay: number) => void] {
  const timer = ref<NodeJS.Timer | null>(null)

  const setTimer = (fn: () => void, delay: number) => {
    timer.value = setTimeout(fn, delay)
  }

  return [timer, setTimer]
}

export function useTriggerEvents(
  state: PopperState,
  visibility: ComputedRef<boolean>,
  timer: Ref<NodeJS.Timer | null>,
  action: { show(): void; hide(): void },
): ComputedRef<PopperTriggerEvents> {
  const triggerEventsMap: Record<PopperTrigger, (keyof PopperTriggerEvents)[]> = {
    click: ['onClick'],
    focus: ['onFocus', 'onBlur'],
    hover: ['onMouseenter', 'onMouseleave'],
    contextmenu: ['onContextmenu'],
    manual: [],
  }
  return computed(() => mapTriggerEvents(triggerEventsMap[state.trigger], state, visibility, timer, action))
}

export function usePopperEvents(
  state: PopperState,
  timer: Ref<NodeJS.Timer | null>,
  hide: () => void,
): ComputedRef<PopperEvents> {
  const onMouseenter = () => {
    if (state.trigger === 'hover' && state.allowEnter && timer.value) {
      clearTimeout(timer.value)
      timer.value = null
    }
  }

  const onMouseleave = () => {
    if (state.trigger !== 'hover') {
      return
    }
    hide()
  }

  return computed(() => ({ onMouseenter, onMouseleave }))
}
