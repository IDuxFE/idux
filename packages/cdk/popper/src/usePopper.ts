import type { Instance } from '@popperjs/core'
import type { PopperOptions, PopperInstance, PopperElement } from './types'

import { watch, WatchStopHandle } from 'vue'
import { createPopper } from '@popperjs/core'
import { isEqual } from 'lodash-es'
import { convertElement } from '@idux/cdk/utils'
import {
  useState,
  useBaseOptions,
  useElement,
  usePlacement,
  usePopperEvents,
  useTimer,
  useTriggerEvents,
  useVisibility,
  useDelay,
} from './hooks'
import { convertOptions } from './utils'

export function usePopper<TE extends PopperElement = PopperElement, PE extends PopperElement = PopperElement>(
  options: PopperOptions = {},
): PopperInstance<TE, PE> {
  let popperInstance: Instance | null = null

  const triggerRef = useElement<TE>()
  const popperRef = useElement<PE>()
  const arrowRef = useElement<HTMLElement>()

  const state = useState(options)
  const baseOptions = useBaseOptions(state)
  const visibility = useVisibility(state)
  const { placement, updatePlacement } = usePlacement(state)
  const delay = useDelay(state)

  const { setTimer, clearTimer } = useTimer()
  const triggerEvents = useTriggerEvents(state, { visibility, show, hide })
  const popperEvents = usePopperEvents(state, { show, hide })

  function toggle(visible: boolean, delay: number): void {
    clearTimer()

    const action = () => {
      state.visible = visible
    }
    if (delay > 0) {
      setTimer(action, delay)
    } else {
      action()
    }
  }

  function show(showDelay = delay.value.show): void {
    toggle(true, showDelay)
  }

  function hide(hideDelay = delay.value.hide): void {
    toggle(false, hideDelay)
  }

  function update(options: Partial<PopperOptions>): void {
    Object.entries(options).forEach(([key, value]) => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      if (value !== undefined && !isEqual(value, (state as any)[key])) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        ;(state as any)[key] = value
      }
    })
  }

  function forceUpdate(): void {
    popperInstance?.forceUpdate()
  }

  function destroy(): void {
    clearTimer()
    if (!popperInstance) {
      return
    }
    popperInstance.destroy()
    popperInstance = null
  }

  let initWatchStop: WatchStopHandle | null = null

  function initialize(): void {
    if (initWatchStop) {
      initWatchStop()
    }

    initWatchStop = watch(
      [triggerRef, popperRef],
      ([trigger, popper]) => {
        const triggerElement = convertElement(trigger)
        const popperElement = convertElement(popper)
        if (!triggerElement || !popperElement) {
          return
        }
        destroy()
        const options = convertOptions(baseOptions.value, { arrowElement: convertElement(arrowRef), updatePlacement })
        popperInstance = createPopper(triggerElement, popperElement, options)
      },
      { immediate: true },
    )
  }

  watch(visibility, value => {
    if (value) {
      clearTimer()
      popperInstance?.update()
    }
  })

  watch([baseOptions, arrowRef], ([currBaseOptions, arrowElement]) => {
    popperInstance?.setOptions(
      convertOptions(currBaseOptions, { arrowElement: convertElement(arrowElement), updatePlacement }),
    )
  })

  return {
    visibility,
    placement,
    triggerRef,
    triggerEvents,
    popperRef,
    popperEvents,
    arrowRef,
    initialize,
    show,
    hide,
    update,
    forceUpdate,
    destroy,
  }
}
