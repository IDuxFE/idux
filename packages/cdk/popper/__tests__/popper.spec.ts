import { onBeforeUnmount, onMounted } from 'vue'
import { flushPromises, mount } from '@vue/test-utils'
import { wait } from '@tests'
import { usePopper } from '../src/usePopper'
import { PopperOptions } from '../src/types'

describe('usePopper', () => {
  const PopperTestMount = (options?: PopperOptions) =>
    mount({
      setup() {
        const instance = usePopper(options)
        onMounted(() => instance.initialize())
        onBeforeUnmount(() => instance.destroy())
        return instance
      },
      template: `
      <div id="trigger" ref="triggerRef" v-bind="triggerEvents">Trigger</div>
      <div v-if="visibility" id="overlay" ref="popperRef" v-bind="popperEvents">Popper</div>
    `,
    })

  describe('options', () => {
    test('allowEnter work', async () => {
      const wrapper = PopperTestMount({ allowEnter: true, hideDelay: 100 })
      await flushPromises()
      await wrapper.find('#trigger').trigger('mouseenter')

      expect(wrapper.find('#overlay').exists()).toBe(true)

      await wrapper.find('#trigger').trigger('mouseleave')
      await wrapper.find('#overlay').trigger('mouseenter')
      await wait(100)

      expect(wrapper.find('#overlay').exists()).toBe(true)

      await wrapper.find('#overlay').trigger('mouseleave')
      await wait(100)

      expect(wrapper.find('#overlay').exists()).toBe(false)

      // false
      wrapper.vm.update({ allowEnter: false })
      await flushPromises()
      await wrapper.find('#trigger').trigger('mouseenter')

      expect(wrapper.find('#overlay').exists()).toBe(true)

      await wrapper.find('#trigger').trigger('mouseleave')
      await wrapper.find('#overlay').trigger('mouseenter')
      await wait(100)

      expect(wrapper.find('#overlay').exists()).toBe(false)
    })

    test('autoAdjust work', async () => {
      const wrapper = PopperTestMount({ autoAdjust: true })
      await flushPromises()

      expect(wrapper.html()).toMatchSnapshot()

      wrapper.vm.update({ autoAdjust: false })
      await flushPromises()

      expect(wrapper.html()).toMatchSnapshot()
    })

    test('disabled work', async () => {
      const wrapper = PopperTestMount({ disabled: true, visible: true })
      await flushPromises()

      expect(wrapper.find('#overlay').exists()).toBe(false)

      wrapper.vm.update({ disabled: false })
      await flushPromises()

      expect(wrapper.find('#overlay').exists()).toBe(true)
    })

    test('offset work', async () => {
      const wrapper = PopperTestMount({ offset: [4, 4], visible: true })
      await flushPromises()

      expect(wrapper.html()).toMatchSnapshot()

      wrapper.vm.update({ offset: [8, 8] })

      await flushPromises()

      expect(wrapper.html()).toMatchSnapshot()
    })

    test('placement work', async () => {
      const wrapper = PopperTestMount({ visible: true })
      const placements = [
        'topStart',
        'top',
        'topEnd',
        'rightStart',
        'right',
        'rightEnd',
        'bottomStart',
        'bottom',
        'bottomEnd',
        'leftStart',
        'left',
        'leftEnd',
      ] as const

      await Promise.all(
        placements.map(async placement => {
          wrapper.vm.update({ placement })

          await flushPromises()

          expect(wrapper.html()).toMatchSnapshot()
        }),
      )
    })

    test('trigger', async () => {
      const wrapper = PopperTestMount()

      // hover

      await flushPromises()
      await wrapper.find('#trigger').trigger('mouseenter')
      expect(wrapper.find('#overlay').exists()).toBe(true)
      await wrapper.find('#trigger').trigger('mouseleave')
      expect(wrapper.find('#overlay').exists()).toBe(false)

      // focus
      wrapper.vm.update({ trigger: 'focus' })

      await flushPromises()
      await wrapper.find('#trigger').trigger('focus')
      expect(wrapper.find('#overlay').exists()).toBe(true)
      await wrapper.find('#trigger').trigger('blur')
      expect(wrapper.find('#overlay').exists()).toBe(false)

      // click
      wrapper.vm.update({ trigger: 'click' })

      await flushPromises()
      await wrapper.find('#trigger').trigger('click')
      expect(wrapper.find('#overlay').exists()).toBe(true)
      await wrapper.find('#trigger').trigger('click')
      expect(wrapper.find('#overlay').exists()).toBe(false)

      // contextmenu
      wrapper.vm.update({ trigger: 'contextmenu' })

      await flushPromises()
      await wrapper.find('#trigger').trigger('contextmenu')
      expect(wrapper.find('#overlay').exists()).toBe(true)
      await wrapper.find('#trigger').trigger('contextmenu')
      expect(wrapper.find('#overlay').exists()).toBe(true)
      await wrapper.find('#trigger').trigger('click')
      expect(wrapper.find('#overlay').exists()).toBe(false)

      wrapper.vm.update({ trigger: 'manual' })
      await flushPromises()

      expect(Object.keys(wrapper.vm.triggerEvents).length).toEqual(0)
      expect(Object.keys(wrapper.vm.popperEvents).length).toEqual(0)
    })

    test('visible work', async () => {
      const wrapper = PopperTestMount({ visible: true })
      await flushPromises()

      expect(wrapper.find('#overlay').exists()).toBe(true)

      wrapper.vm.update({ visible: false })
      await flushPromises()

      expect(wrapper.find('#overlay').exists()).toBe(false)
    })

    test('hideDelay work', async () => {
      const wrapper = PopperTestMount({ visible: true, hideDelay: 100 })
      await wrapper.find('#trigger').trigger('mouseleave')
      await flushPromises()

      expect(wrapper.find('#overlay').exists()).toBe(true)

      await wait(100)

      expect(wrapper.find('#overlay').exists()).toBe(false)
    })

    test('showDelay work', async () => {
      const wrapper = PopperTestMount({ trigger: 'hover', showDelay: 100 })
      await wrapper.find('#trigger').trigger('mouseenter')
      await flushPromises()

      expect(wrapper.find('#overlay').exists()).toBe(false)

      await wait(100)

      expect(wrapper.find('#overlay').exists()).toBe(true)
    })
  })

  // TODO
  describe('instance', () => {})
})
