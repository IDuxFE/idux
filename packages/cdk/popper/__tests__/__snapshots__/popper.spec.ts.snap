// Vitest Snapshot v1

exports[`usePopper > options > autoAdjust work 1`] = `
"<div id=\\"trigger\\">Trigger</div>
<!--v-if-->"
`;

exports[`usePopper > options > autoAdjust work 2`] = `
"<div id=\\"trigger\\">Trigger</div>
<!--v-if-->"
`;

exports[`usePopper > options > offset work 1`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, -4px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"top\\">Popper</div>"
`;

exports[`usePopper > options > offset work 2`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, -8px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"top\\">Popper</div>"
`;

exports[`usePopper > options > placement work 1`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 2`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 3`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 4`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 5`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 6`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 7`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 8`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 9`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 10`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 11`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;

exports[`usePopper > options > placement work 12`] = `
"<div id=\\"trigger\\">Trigger</div>
<div id=\\"overlay\\" style=\\"position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" data-popper-reference-hidden=\\"\\" data-popper-escaped=\\"\\" data-popper-placement=\\"left-end\\">Popper</div>"
`;
